Sef zfs quota

kithrup opened the pull request:
For ticket #3335.

Some notes:

- This doesn't change quota, edquota, or others.  ZFS quotas are managed via the zfs command still.
- rpc.rquotad needs to be running on the server for it to be queryable by the client.
- Creating the PR for review and testing purposes.

kithrup committed:
Re-apply changes to enable remote quota checking for ZFS datasets.

kithrup committed:
Re-apply changes to enable remote quota checking for ZFS datasets.

Commit:
Some minor tweaks to make the quota(8) command on FreeBSD print
out better results.

Commit:
Some minor tweaks to make the quota(8) command on FreeBSD print
out better results.

Commit:
Start cleaning up the code.

Commit:
Start cleaning up the code.

Commit:
And some more cleanup.

Commit:
And some more cleanup.

kithrup commented:
Jenkins: retest this please

amotin commented on lib/libutil/quotafile.c:
```diff
@@ -118,8 +118,6 @@ quota_open(struct fstab *fs, int quotatype, int openflags)
 	int qcmd, serrno;
 
-	if (strcmp(fs->fs_vfstype, "ufs"))
-		return (NULL);
```
I think instead of removing this, it should be moved down after quotactl() call, since code after that is indeed UFS-specific.  Plus also hasquota() I would probably move after that, since that call makes no sense if quotactl() succeed.

amotin commented on include/rpcsvc/rquota.x:
```diff
@@ -37,7 +72,7 @@ struct rquota {
 enum gqr_status {
 	Q_OK = 1,		/* quota returned */
-	Q_NOQUOTA = 2,  	/* noquota for uid */
+	Q_NOQUOTA = 2, 		/* noquota for uid */
```
Are the changes to this file were taken from somewhere exactly?  Because here you removed one space, but left another one.  Same as below there are few extra spaces between tabs.

kithrup replied:
They were applied by a patch, and then I used a tab to align them.  And for some reason didn't check the previous character when I did so.

amotin commented on include/rpcsvc/rquota.x:
```diff
@@ -63,5 +107,42 @@ program RQUOTAPROG {
+	 	 * Get active quotas only
+		 */
+		setquota_rslt
+		RQUOTAPROC_SETACTIVEQUOTA(setquota_args) = 4;
```
The method is "SET", but comment tells "Get".

amotin commented on libexec/rpc.rquotad/rquotad.c:
```diff
@@ -172,51 +205,81 @@ sendquota(struct svc_req *request, SVCXPRT *transp)
+	 * compiles with -Werrr,-Wcast-qual it fails.
+	 */
+	static char userquota[] = "userquota";
+	struct fstab fst = { .fs_file = path, .fs_mntops=userquota };
```
This looks like a hack.  I suppose it may work for ZFS, which is never in fstab and always have quotas enabled, but for UFS I guess it may at least change behavior if real fstab does not have any quota options, while here you are hardcoding one.

kithrup replied:
The particular layout is a hack, because of compiler options not allowing it to be written a sane way.  However, other than that, setting up the fst structure there is the right way:  this is rquotad, and it's being asked about a filesystem.  We can assume that it's mounted on the remote end, and that means it has to be mounted on the server end as well.  The quota_open() call will fail if quotas are not enabled on the mount point, so it all works out okay.

kithrup replied:
I've gone through it again.  It does result in different behaviour, but I am going to maintain that the old behaviour -- which only looks at fstab, not mnttab -- is wrong.  Now, that gets to a question of whether zfs should add a generated "userquotas" and/or "groupquotas" for the mnttops, and ... that's a pretty good question.  Then it would instead be a matter of getting the mount options for a path at runtime, and determining whether it's set there.

kithrup replied:
This needs review/comment still.  No response to my comment above, which was an implied question.

Options are to stall this until I make zfs mounting set userquotas and groupquotas, or accept it into freenas but not push it to freebsd, or to accept the change in behaviour and try to push it to freebsd.

amotin commented on usr.bin/quota/quota.c:
```diff
@@ -566,21 +566,17 @@ getufsquota(struct fstab *fs, struct quotause *qup, long id, int quotatype)
 	char *cp, host[NI_MAXHOST];
-
+	enum clnt_stat call_stat;
+	
```
You are again adding extra tab.

amotin commented on usr.bin/quota/quota.c:
```diff
@@ -602,11 +598,22 @@ getnfsquota(struct statfs *fst, struct quotause *qup, long id, int quotatype)
+			      RQUOTAPROC_GETQUOTA, (xdrproc_t)xdr_ext_getquota_args, (char *)&gq_args,
+			      (xdrproc_t)xdr_getquota_rslt, (char *)&gq_rslt);
+	if (call_stat == RPC_PROGVERSMISMATCH) {
+		warnx("%s(%d): Oops, trying older version", __FUNCTION__, __LINE__);
```
This looks like a debugging message rather then production code.

amotin commented on usr.bin/quota/quota.c:
```diff
@@ -669,8 +676,9 @@ callaurpc(char *host, int prognum, int versnum, int procnum,
- 
-	return ((int) clnt_stat);
+	if (clnt_stat != 0)
+		warnx("%s(%d): clnt_stat = %d", __FUNCTION__, __LINE__, (int)clnt_stat);
```
Again, I am not sure it should be logged every time.

amotin reviewed (commented):
I see this viable, but have some comments.

kithrup committed:
Feedback from Alexander.

kithrup committed:
Feedback from Alexander.

kithrup commented:
I just pushed changes based on Alexander's feedback.

amotin commented on usr.bin/quota/quota.c:
```diff
@@ -602,11 +598,21 @@ getnfsquota(struct statfs *fst, struct quotause *qup, long id, int quotatype)
+			      RQUOTAPROC_GETQUOTA, (xdrproc_t)xdr_ext_getquota_args, (char *)&gq_args,
+			      (xdrproc_t)xdr_getquota_rslt, (char *)&gq_rslt);
+	if (call_stat == RPC_PROGVERSMISMATCH) {
+		old_gq_args.gqa_pathp = cp + 1;
```
Shouldn't (quotatype == USRQUOTA) be checked here?  As I understand, old RPC does not support group quotas.

kithrup replied:
Done, sorry I missed that earlier.

amotin reviewed (commented):
One more comment, while otherwise looks good to me.  Just may be some style could be tuned, for example, blank lines before multi-line comments and after blocks they comment.

kithrup committed:
A missed feedback from Alexander which he reminded me of:
the old RPC functions don't support group quotas.

kithrup committed:
A missed feedback from Alexander which he reminded me of:
the old RPC functions don't support group quotas.

amotin commented on usr.bin/quota/quota.c:
```diff
@@ -602,11 +598,26 @@ getnfsquota(struct statfs *fst, struct quotause *qup, long id, int quotatype)
+			old_gq_args.gqa_pathp = cp + 1;
+			old_gq_args.gqa_uid = id;
+			call_stat = callaurpc(host, RQUOTAPROG, RQUOTAVERS,
+					      RQUOTAPROC_GETQUOTA, (xdrproc_t)xdr_getquota_args, (char *)&old_gq_args,
```
I would invert the quotatype check to reduce the code indentation, since it obviously does not fit into 80 columns now.

amotin reviewed (commented):
Thanks, but I'd modify it.

Commit:
Merge branch 'freenas/11-stable' into sef-zfs-quota

Commit:
Change rpc.rquotad to be a bit less hacky, and also a bit safer.
This is based on looking at the quota_* routines in libutil --
we only care about filesystems that are mounted, and that supporot
the quoota system calls (which is UFS and ZFS).

Commit:
Change rpc.rquotad to be a bit less hacky, and also a bit safer.
This is based on looking at the quota_* routines in libutil --
we only care about filesystems that are mounted, and that supporot
the quoota system calls (which is UFS and ZFS).

Commit:
Merge branch 'sef-zfs-quota' of https://github.com/freenas/os into sef-zfs-quota

kithrup commented:
Jenkins: retest this please

Commit:
Merge branch 'freenas/11-stable' into sef-zfs-quota

kmoore134 commented:
retest

kmoore134 commented:
retest

amotin commented:
I suppose this was done long ago. Why do we still have this pull request?
